net: Remove useless asserts

jukkar opened the pull request:
Remove assert calls from networking code where the asserts were used incorrect to try to check something at runtime. As asserts are normally disabled, the checks were pointless. Replace the asserts with proper runtime checks where applicable.

jukkar committed:
net: trickle: Remove useless asserts

Replace asserts by runtime checks as the APIs can be used
from applications.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: vlan: Remove useless assert

The assert is not needed and it was in wrong place anyway.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: websocket: Remove assert call

Remove the useless assert as it is not needed at all here.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: if: Replace asserts by proper runtime checks

The asserts were not proper here, replace those by runtime
checks as the functions can be called from applications and
asserts are not meant for error checking.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented:
I do want to question though, if the _contract  of a function_ dictates that an argument can't be `NULL`, do you want to run the check each time the function is called?

C does not have the ability to force non-NULL pointers like other languages can sadly.

jukkar commented:
> I do want to question though, if the _contract of a function_ dictates that an argument can't be `NULL`, do you want to run the check each time the function is called?

No, we do not need to check argument that cannot be null. I just started to go through network code base to see how we use the asserts as this assert misuse was mentioned yesterday in arch wg meeting.
Some of the changes in this PR are for APIs that could be called from applications in which case using runtime checks is more appropriate. The changes in this PR were somewhat mechanic, although I tried to check whether the runtime check is needed. But I probably missed cases where we already know that the value cannot be null.

rlubos commented on subsys/net/ip/net_context.c:
```diff
@@ -2837,8 +2876,13 @@ static enum net_verdict net_context_raw_packet_received(
 
-	NET_ASSERT(context);
-	NET_ASSERT(net_pkt_iface(pkt));
+	if (context == NULL) {
```
So while I'm ok with and support getting rid of asserts in public APIs (the function should fail, not crash the application), I'm not convinced we should do this for internal (static) functions. Here we have full control of the provided arguments, and providing incorrect ones most likely means we have a bug somewhere in our own code. IMO leaving asserts in such cases makes them easier to identify.

jukkar replied:
Makes sense, I will rework the PR.

jukkar commented:
* Removed changes to net_context, tcp, route and offload code. These can be considered internal APIs and we have already checks in various places before these APIs are called.
* The changes to public `net_if` APIs are still there, but keep asserts to internal APIs

jukkar commented:
@pdgendt PTAL

pdgendt commented:
> @pdgendt PTAL

Sorry, forgot about this.

I do have some reluctance about removing the asserts, these do have a purpose in development IMO.
I agree with the failsafe checks, but I'd keep the asserts too if the function's contract expects a certain parameter (mostly non-NULL).
WDYT?

pdgendt reviewed (approved):
I don't want to block this either.

jukkar commented:
> I do have some reluctance about removing the asserts, these do have a purpose in development IMO.
> I agree with the failsafe checks, but I'd keep the asserts too if the function's contract expects a certain parameter (mostly non-NULL).
> WDYT?

I went through the asserts in the code and came up with this proposal what is now merged. The motivation for the changes in each commits were:
* Trickle: public API calls, we need to check the parameters anyway and the assert is pretty much useless in this case
* VLAN: the assert was after the use of the variable, and the variable could not be null anyway -> removed
* websocket: the assert was useless as the value cannot be null in all the call paths leading to this function
* iface: removed asserts from public api functions, left the asserts to internal functions

Note that in this 2nd version of this PR, I left many of the asserts in place that I removed in 1st version. Perhaps your reluctance was about this earlier version?

pdgendt commented:
I was referring to the merged version, something like the following would perfect sense IMO:

```c
NET_ASSERT(iface != NULL);
if (iface == NULL) {
	return;
}
```
